serve.go

```diff
@@ -0,0 +1,128 @@
+package main
+
+import (
+	"io"
+	"log"
+	"math/rand"
+	"net/http"
+	"strconv"
+
+	"github.com/stripe/sequins/blocks"
+)
+
+// serveKey is the entrypoint for incoming HTTP requests. It looks up the value
+// locally, for, failing that, asks a peer that has it. If the request was
+// already proxied to us, it is not proxied further.
+func (vs *version) serveKey(w http.ResponseWriter, r *http.Request, key string) {
+	// If we don't have any data for this version at all, that's a 404.
+	if vs.numPartitions == 0 {
+		vs.serveNotFound(w)
+	}
+
+	partition, alternatePartition := blocks.KeyPartition(key, vs.numPartitions)
+	bs := vs.getBlockStore()
+	if bs != nil && vs.hasPartition(partition) || vs.hasPartition(alternatePartition) {
+		record, err := bs.Get(key)
+		if err != nil {
+			vs.serveError(w, key, err)
+			return
+		}
+
+		vs.serveLocal(w, key, record)
+	} else if r.URL.Query().Get("proxy") == "" {
+		vs.serveProxied(w, key, partition, alternatePartition)
+	} else {
+		vs.serveError(w, key, errProxiedIncorrectly)
+	}
+}
+
+func (vs *version) serveLocal(w http.ResponseWriter, key string, record *blocks.Record) {
+	if record == nil {
+		vs.serveNotFound(w)
+		return
+	}
+
+	defer record.Close()
+	w.Header().Set(versionHeader, vs.name)
+	w.Header().Set("Content-Length", strconv.FormatUint(record.ValueLen, 10))
+	w.Header().Set("Last-Modified", vs.created.UTC().Format(http.TimeFormat))
+	_, err := io.Copy(w, record)
+	if err != nil {
+		// We already wrote a 200 OK, so not much we can do here except log.
+		log.Printf("Error streaming response for /%s/%s (version %s): %s", vs.db, key, vs.name)
+	}
+}
+
+func (vs *version) serveProxied(w http.ResponseWriter, key string, partition, alternatePartition int) {
+	// Shuffle the peers, so we try them in a random order.
+	// TODO: We don't want to blacklist nodes, but we can weight them lower
+	peers := shuffle(vs.partitions.getPeers(partition))
+	if len(peers) == 0 {
+		log.Printf("No peers available for /%s/%s (version %s)", vs.db, key, vs.name)
+		w.WriteHeader(http.StatusBadGateway)
+		return
+	}
+
+	resp, peer, err := vs.proxy(peers)
+	if err == nil && resp.StatusCode == 404 && alternatePartition != partition {
+		log.Println("Trying alternate partition for pathological key", key)
```

Comment:
Read [this comment](https://github.com/stripe/sequins/blob/1.0/blocks/hashcode.go#L31-L78) if you really want to know.